### PR TITLE
feat(react-core): add useDataState

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes

Add `useDataState` react utility hook

#### Issue #, if available
NA

#### Description of how you validated changes
`yarn test`

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-core/src/hooks/useDataState.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  message: string | undefined;
+}
+
+const getActionState = <T>(data: T): ActionState<T> => ({
+  data,
+  isLoading: false,
+  message: undefined,
+});
+
+export default function useDataState<T, K>(
```
Comment (reesscot) [inline packages/react-core/src/hooks/useDataState.ts:15]: Probably be good to put a JSDoc comment here on example usage

**Review by reesscot — APPROVED** [review]

File: packages/react-core/src/hooks/index.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+export { default as useDataState } from './useDataState';
```
Comment (reesscot) [inline packages/react-core/src/hooks/index.ts:1]: Is this exported publicly?

Comment (calebpollman) [reply packages/react-core/src/hooks/index.ts:1]: Nope

File: packages/react-core/src/hooks/useDataState.ts (hunk 1 above)
Comment (calebpollman) [reply packages/react-core/src/hooks/useDataState.ts:15]: Will add later 👍

**Review by zchenwei — APPROVED** [review]
Nice!

**Pull request merged**